File: server/review/app/models/review.py
from sqlalchemy import Column, String, Float, Integer, Boolean, DateTime, Text, ForeignKey
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid

Base = declarative_base()

class Review(Base):
    __tablename__ = "reviews"
    
    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    productId = Column(String(36), nullable=False, index=True)
    userId = Column(String(36), nullable=False, index=True)
    rating = Column(Float, nullable=False)
    title = Column(String(200))
    content = Column(Text)
    helpfulVotes = Column(Integer, default=0)
    verifiedPurchase = Column(Boolean, default=False)
    sentiment = Column(String(20))  # LIKE, NEUTRAL, DISLIKE
    level = Column(Integer, default=1)
    parentId = Column(String(36), ForeignKey('reviews.id'), nullable=True)
    reviewDate = Column(DateTime, default=datetime.now)
    createdAt = Column(DateTime, default=datetime.now)
    updatedAt = Column(DateTime, default=datetime.now, onupdate=datetime.now)
    
    # Self-referential relationship for replies
    replies = relationship("Review", backref="parent", remote_side=[id])

File: server/product/src/main/java/com/ecommerce/product/entity/Review.java
package com.ecommerce.product.entity;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "reviews")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Review {

    @Id
    @GeneratedValue(generator = "UUID")
    @org.hibernate.annotations.UuidGenerator
    @Column(name = "id", updatable = false, nullable = false, columnDefinition = "VARCHAR(36)")
    private String id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id", nullable = false)
    private Product product;

    @Column(name = "user_id")
    private String userId;

    @Column(nullable = false, precision = 2, scale = 1)
    private BigDecimal rating;

    private String title;

    @Column(columnDefinition = "TEXT")
    private String content;

    @Column(name = "helpful_votes")
    private Integer helpfulVotes;

    @Column(name = "verified_purchase")
    private Boolean verifiedPurchase;

    @Column(name = "review_date")
    private LocalDateTime reviewDate;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        if (id == null) {
            id = UUID.randomUUID().toString();
        }
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (reviewDate == null) {
            reviewDate = LocalDateTime.now();
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}